pilot/tools/debug/pilot_cli.go

```diff
@@ -114,12 +126,26 @@ func NewPodInfo(nameOrAppLabel string, kubeconfig string) *PodInfo {
 				IP:        pod.Status.PodIP,
 			}
 		}
+		if istio, ok := pod.ObjectMeta.Labels["istio"]; ok && istio == nameOrAppLabel {
+			log.Infof("Found pod %s.%s~%s matching app label %q", pod.Name, pod.Namespace, pod.Status.PodIP, nameOrAppLabel)
+			return &PodInfo{
+				Name:      pod.Name,
+				Namespace: pod.Namespace,
+				IP:        pod.Status.PodIP,
+			}
+		}
 	}
 	log.Warnf("Cannot find pod with name or app label matching %q in registry.", nameOrAppLabel)
 	return nil
 }
 
 func (p PodInfo) makeNodeID() string {
+	if strings.HasPrefix(p.Name, "istio-ingressgateway") || strings.HasPrefix(p.Name, "istio-egressgateway") {
+		return fmt.Sprintf("router~%s~%s.%s~%s.svc.cluster.local", p.IP, p.Name, p.Namespace, p.Namespace)
```

Comment:
I think there's a helper function in the model package that does this. We should probably use that to pick up any future changes,

Reply:
I cannot find the helper function that fits the need. I think pilot figures out node type based on discovery request params instead of pod info?

Reply:
https://github.com/istio/istio/blob/master/pilot/pkg/model/context.go#L98

Reply:
Thats I good point it's not really consumable as a helper function, we should probably have one though. Maybe split out that function into two with one being a standalone string conversion that can be consumed anywhere?

This might be overkill but it will prevent future breaking changes. ¯\_(ツ)_/¯

Reply:
How do you split that function into two? What kind of string conversion are you talking about? Pilot-agent does not derive node type from pod name. The functionalities are totally different here in this tool and in pilot-agent.

The new code is just for convenience. It won't break anything and does no worse than it used to be when some new changes introduced. I really don't want to drag the main code base for the need here. Beside this is just a tool used by istio developers, any new cases could be fixed easily by us.

Reply:
You're right its probably overkill. I will LGTM it but my review has no power. 🙂